File: internal/parser/v3/parser.go
// Package parser implements a flat-node-pool PDXScript parser (Option A).
//
// All nodes live in a single []Node slice — no heap pointers inside Node.
// Parent-child relationships are expressed via an index array: each node
// stores a range [ChildStart, ChildEnd) into Tree.Index, and Tree.Index[i]
// is the index of the i-th child in Tree.Nodes.
//
// String values are byte-offset ranges into the original source slice,
// so no string copies are made during parsing.
//
// The parser is error-tolerant: it accumulates diagnostics and produces a
// partial tree rather than stopping at the first error. Callers should check
// len(diags) > 0 rather than treating a non-nil tree as fully valid.
//
// This is the fastest variant. For the pointer-based tree see internal/parser/v2.
// For the participle-based reference see internal/parser/v1.
package v3

import (
	"fmt"

	"pdxl/internal/lexer"
)

// ── Node kinds ────────────────────────────────────────────────────────────────

type NodeKind uint8

const (
	KindFile        NodeKind = iota // root; children = top-level items
	KindField                       // children[0]=key scalar, children[1]=value; Op set
	KindBlock                       // children = items
	KindTaggedBlock                 // children = items; SrcStart..SrcEnd = tag text
	KindScalar                      // leaf; SrcStart..SrcEnd is the text
)

// ── Node ──────────────────────────────────────────────────────────────────────

// Node is a pointer-free AST node.
type Node struct {
	Kind NodeKind

	// Source byte range (for scalars, field keys, tagged-block tag names).
	SrcStart uint32
	SrcEnd   uint32

	// Op stores the operator tag for Field nodes.
	Op lexer.Tag

	// Child range into Tree.Index (not Tree.Nodes).
	ChildStart uint32
	ChildEnd   uint32
}

// Value returns the source text for this node.
func (n Node) Value(src []byte) string {
	return string(src[n.SrcStart:n.SrcEnd])
}

// OpString returns the operator as its source symbol (e.g. "=", "?=", ">=").
func (n Node) OpString() string {
	switch n.Op {
	case lexer.TagEqual:
		return "="
	case lexer.TagEqualEqual:
		return "=="
	case lexer.TagNotEqual:
		return "!="
	case lexer.TagQuestionEqual:
		return "?="
	case lexer.TagGreaterThan:
		return ">"
	case lexer.TagGreaterEqual:
		return ">="
	case lexer.TagLessThan:
		return "<"
	case lexer.TagLessEqual:
		return "<="
	}
	return n.Op.String()
}

// ── Tree ──────────────────────────────────────────────────────────────────────

// Tree is the result of parsing.
// Nodes[0] is always the KindFile root.
// Index provides child indirection: node.Children are Index[ChildStart:ChildEnd],
// each element being an index into Nodes.
type Tree struct {
	Nodes []Node
	Index []uint32 // child index array
	Src   []byte
}

// Root returns the file root node.
func (t *Tree) Root() Node { return t.Nodes[0] }

// ChildRefs returns the child indices of n into Tree.Nodes without allocating.
// Prefer this in hot paths; use Children for convenience.
func (t *Tree) ChildRefs(n Node) []uint32 {
	return t.Index[n.ChildStart:n.ChildEnd]
}

// Children returns the direct children of n as Node values.
func (t *Tree) Children(n Node) []Node {
	refs := t.ChildRefs(n)
	out := make([]Node, len(refs))
	for i, idx := range refs {
		out[i] = t.Nodes[idx]
	}
	return out
}

// ── Diagnostics ───────────────────────────────────────────────────────────────

// Severity classifies how serious a diagnostic is.
type Severity uint8

const (
	SeverityError   Severity = iota
	SeverityWarning
)

// Diagnostic is a parse problem with a source location.
// Offset is a byte offset into the source; use lexer.Token{Start: d.Offset}.FormatPosition
// to convert to a line:column string.
type Diagnostic struct {
	Filename string
	Offset   int
	Msg      string
	Severity Severity
}

func (d Diagnostic) String() string {
	tok := lexer.Token{Start: d.Offset, End: d.Offset}
	return fmt.Sprintf("%s: %s", tok.FormatPosition(d.Filename, nil), d.Msg)
}

// ── parser ────────────────────────────────────────────────────────────────────

// invalidIdx is returned by parse functions when no node was produced.
const invalidIdx = ^uint32(0)

type parser struct {
	tokens   []lexer.Token
	src      []byte
	filename string
	pos      int    // current index into tokens
	nodes    []Node // flat node pool
	index    []uint32
	diags    []Diagnostic
}

func newParser(filename string, src []byte) *parser {
	tokens := lexer.Tokenize(src)
	cap := len(tokens) / 2
	return &parser{
		tokens:   tokens,
		src:      src,
		filename: filename,
		nodes:    make([]Node, 0, cap),
		index:    make([]uint32, 0, cap),
	}
}

// allocNode appends a node and returns its index.
func (p *parser) allocNode(n Node) uint32 {
	idx := uint32(len(p.nodes))
	p.nodes = append(p.nodes, n)
	return idx
}

// peek returns the tag at pos+offset.
func (p *parser) peek(offset int) lexer.Tag {
	i := p.pos + offset
	if i >= len(p.tokens) {
		return lexer.TagEOF
	}
	return p.tokens[i].Tag
}

// advance consumes and returns the current token.
func (p *parser) advance() lexer.Token {
	if p.pos >= len(p.tokens) {
		return lexer.Token{Tag: lexer.TagEOF}
	}
	tok := p.tokens[p.pos]
	p.pos++
	return tok
}

func (p *parser) currentOffset() int {
	if p.pos < len(p.tokens) {
		return p.tokens[p.pos].Start
	}
	if len(p.tokens) > 0 {
		return p.tokens[len(p.tokens)-1].End
	}
	return 0
}

func (p *parser) addDiag(offset int, severity Severity, msg string) {
	p.diags = append(p.diags, Diagnostic{
		Filename: p.filename,
		Offset:   offset,
		Msg:      msg,
		Severity: severity,
	})
}

// synchronize skips tokens until a safe resumption point:
// a closing '}', the start of a plausible new item, or EOF.
// It does NOT consume the token it stops at.
func (p *parser) synchronize() {
	for {
		switch p.peek(0) {
		case lexer.TagEOF, lexer.TagRBrace:
			return
		}
		t0 := p.peek(0)
		if isAtom(t0) {
			t1 := p.peek(1)
			if isOperator(t1) || t1 == lexer.TagLBrace {
				return
			}
		}
		p.advance()
	}
}

func isAtom(tag lexer.Tag) bool {
	switch tag {
	case lexer.TagIdentifier, lexer.TagLiteralNumber, lexer.TagLiteralString,
		lexer.TagLiteralBoolean, lexer.TagMinus, lexer.TagMacroParam,
		lexer.TagScriptValue, lexer.TagScriptMath:
		return true
	}
	return false
}

func isOperator(tag lexer.Tag) bool {
	switch tag {
	case lexer.TagEqual, lexer.TagEqualEqual, lexer.TagNotEqual,
		lexer.TagQuestionEqual, lexer.TagGreaterThan, lexer.TagGreaterEqual,
		lexer.TagLessThan, lexer.TagLessEqual:
		return true
	}
	return false
}

func bindingPower(tag lexer.Tag) int {
	switch tag {
	case lexer.TagColon, lexer.TagDot, lexer.TagPipe:
		return 80
	}
	return 0
}

func (p *parser) peekOpAfterKey() bool {
	i := 1
	for {
		tag := p.peek(i)
		if tag == lexer.TagColon || tag == lexer.TagDot {
			i += 2
			continue
		}
		return isOperator(tag)
	}
}

// ── Parsing ───────────────────────────────────────────────────────────────────

// parseFile parses top-level items and returns the root node index.
func (p *parser) parseFile() uint32 {
	rootIdx := p.allocNode(Node{Kind: KindFile})
	var childIdxs []uint32

	for p.peek(0) != lexer.TagEOF {
		idx := p.parseItem()
		if idx != invalidIdx {
			childIdxs = append(childIdxs, idx)
		}
	}

	start := uint32(len(p.index))
	p.index = append(p.index, childIdxs...)
	p.nodes[rootIdx].ChildStart = start
	p.nodes[rootIdx].ChildEnd = uint32(len(p.index))
	return rootIdx
}

// parseItem returns the index of the parsed node, or invalidIdx for skipped tokens.
func (p *parser) parseItem() uint32 {
	t0 := p.peek(0)
	if t0 == lexer.TagEOF {
		return invalidIdx
	}
	if t0 == lexer.TagRBrace || t0 == lexer.TagRBracket {
		p.advance()
		return invalidIdx
	}
	if t0 != lexer.TagMinus && isAtom(t0) && p.peekOpAfterKey() {
		return p.parseField()
	}
	return p.parseValue(0)
}

// parseField returns the index of a KindField node, or invalidIdx on error.
func (p *parser) parseField() uint32 {
	keyStart := uint32(p.tokens[p.pos].Start)
	keyEnd := uint32(p.tokens[p.pos].End)
	p.advance()
	for p.peek(0) == lexer.TagColon || p.peek(0) == lexer.TagDot {
		p.advance() // connector
		if p.pos < len(p.tokens) {
			keyEnd = uint32(p.tokens[p.pos].End)
			p.advance()
		}
	}

	if !isOperator(p.peek(0)) {
		p.addDiag(p.currentOffset(), SeverityError,
			fmt.Sprintf("expected operator, got %s", p.peek(0)))
		p.synchronize()
		return invalidIdx
	}
	opTok := p.advance()

	keyIdx := p.allocNode(Node{Kind: KindScalar, SrcStart: keyStart, SrcEnd: keyEnd})

	valIdx := p.parseValue(0)
	if valIdx == invalidIdx {
		return invalidIdx
	}

	idxStart := uint32(len(p.index))
	p.index = append(p.index, keyIdx, valIdx)

	return p.allocNode(Node{
		Kind:       KindField,
		SrcStart:   keyStart,
		SrcEnd:     keyEnd,
		Op:         opTok.Tag,
		ChildStart: idxStart,
		ChildEnd:   idxStart + 2,
	})
}

// parseBlockItems parses items until '}' or EOF.
// lbrace is the opening '{' token, used for unclosed-block diagnostics.
func (p *parser) parseBlockItems(lbrace lexer.Token) []uint32 {
	var items []uint32
	for p.peek(0) != lexer.TagRBrace && p.peek(0) != lexer.TagEOF {
		idx := p.parseItem()
		if idx != invalidIdx {
			items = append(items, idx)
		}
	}
	if p.peek(0) == lexer.TagRBrace {
		p.advance()
	} else {
		// EOF reached without closing brace. Note: an inner unclosed block may
		// have consumed the closing '}' intended for this block, so the real
		// mistake may be inside rather than here.
		p.addDiag(int(lbrace.Start), SeverityError, "unclosed block (missing '}'; an inner block may have stolen the closing brace)")
	}
	return items
}

// parseValue returns the index of a value node, or invalidIdx on error.
func (p *parser) parseValue(minBP int) uint32 {
	// Unary minus.
	if p.peek(0) == lexer.TagMinus {
		start := uint32(p.tokens[p.pos].Start)
		p.advance()
		if p.pos >= len(p.tokens) {
			p.addDiag(p.currentOffset(), SeverityError, "unexpected EOF after '-'")
			return invalidIdx
		}
		end := uint32(p.tokens[p.pos].End)
		p.advance()
		for bindingPower(p.peek(0)) > minBP {
			p.advance()
			if p.pos < len(p.tokens) {
				end = uint32(p.tokens[p.pos].End)
				p.advance()
			}
		}
		return p.allocNode(Node{Kind: KindScalar, SrcStart: start, SrcEnd: end})
	}

	// Tagged block.
	if p.peek(0) == lexer.TagIdentifier && p.peek(1) == lexer.TagLBrace {
		tagTok := p.advance()
		lbrace := p.advance() // consume '{'
		items := p.parseBlockItems(lbrace)
		idxStart := uint32(len(p.index))
		p.index = append(p.index, items...)
		return p.allocNode(Node{
			Kind:       KindTaggedBlock,
			SrcStart:   uint32(tagTok.Start),
			SrcEnd:     uint32(tagTok.End),
			ChildStart: idxStart,
			ChildEnd:   uint32(len(p.index)),
		})
	}

	// Plain block.
	if p.peek(0) == lexer.TagLBrace {
		lbrace := p.advance() // consume '{'
		items := p.parseBlockItems(lbrace)
		idxStart := uint32(len(p.index))
		p.index = append(p.index, items...)
		return p.allocNode(Node{
			Kind:       KindBlock,
			ChildStart: idxStart,
			ChildEnd:   uint32(len(p.index)),
		})
	}

	// Atom + optional scope-chain infix.
	if !isAtom(p.peek(0)) {
		p.addDiag(p.currentOffset(), SeverityError,
			fmt.Sprintf("expected value, got %s", p.peek(0)))
		p.synchronize()
		return invalidIdx
	}
	start := uint32(p.tokens[p.pos].Start)
	end := uint32(p.tokens[p.pos].End)
	p.advance()
	for bindingPower(p.peek(0)) > minBP {
		p.advance() // connector
		if p.pos < len(p.tokens) {
			end = uint32(p.tokens[p.pos].End)
			p.advance()
		}
	}
	return p.allocNode(Node{Kind: KindScalar, SrcStart: start, SrcEnd: end})
}

// ── Public API ────────────────────────────────────────────────────────────────

// Parse parses src and returns a Tree along with any diagnostics.
// The tree is always non-nil; a non-empty diagnostics slice means errors were
// found but parsing continued as far as possible.
func Parse(filename string, src []byte) (*Tree, []Diagnostic) {
	p := newParser(filename, src)
	p.parseFile()
	return &Tree{Nodes: p.nodes, Index: p.index, Src: src}, p.diags
}

File: rust/crates/pdxl-parser/src/parser.rs
//! Recursive-descent + Pratt parser, a direct port of `internal/parser/v3`.
//!
//! The algorithm, node-allocation order, and child-index ordering match the Go
//! parser exactly — these are differential-parity targets, not just "equivalent
//! trees". The parser is error-tolerant: it accumulates diagnostics and always
//! returns a (possibly partial) tree.

use std::sync::Arc;

use pdxl_lexer::{Token, TokenKind, tokenize};
use pdxl_src::TextRange;

use crate::diagnostic::{Diagnostic, Parse, Severity};
use pdxl_ast::{Node, NodeId, NodeKind, SyntaxTree};

/// Parses `source`, returning a tree and diagnostics.
///
/// The tree is always produced; check [`Parse::diagnostics`] rather than trusting
/// the tree blindly. `filename` is shared into each diagnostic.
pub fn parse(filename: impl Into<Arc<str>>, source: impl Into<Arc<[u8]>>) -> Parse {
    parse_inner(filename.into(), source.into(), false)
}

/// Parses `source` as an interface script (`.gui`). Identical to [`parse`]
/// except that an unquoted datafunction — `enabled = [ArmyWindow.CanMerge]` —
/// is accepted as a scalar value covering the whole `[…]` text. The script
/// grammar is untouched (it is a Go-parity target); `.gui`'s other extras
/// (`template NAME { }`, `type x = base { }`, `block "n" { }`) already parse
/// as sibling scalars + tagged blocks, mirroring the typed-definition shape.
pub fn parse_gui(filename: impl Into<Arc<str>>, source: impl Into<Arc<[u8]>>) -> Parse {
    parse_inner(filename.into(), source.into(), true)
}

fn parse_inner(filename: Arc<str>, source: Arc<[u8]>, gui: bool) -> Parse {
    let mut p = Parser::new(filename, &source);
    p.gui = gui;
    p.parse_file();
    let tree = SyntaxTree::from_parts(
        source,
        p.nodes.into_boxed_slice(),
        p.index.into_boxed_slice(),
    );
    Parse {
        tree,
        diagnostics: p.diags,
    }
}

struct Parser {
    tokens: Vec<Token>,
    filename: Arc<str>,
    pos: usize,
    nodes: Vec<Node>,
    index: Vec<NodeId>,
    diags: Vec<Diagnostic>,
    /// Interface-script dialect: accept `[Datafunction.Chain]` values.
    gui: bool,
}

/// A blank node template; fields are overwritten by each `alloc_node` caller.
const fn blank(kind: NodeKind) -> Node {
    Node {
        kind,
        range: TextRange::new(0, 0),
        operator: TokenKind::Invalid,
        child_start: 0,
        child_end: 0,
    }
}

impl Parser {
    fn new(filename: Arc<str>, src: &[u8]) -> Self {
        let tokens = tokenize(src);
        let cap = tokens.len() / 2;
        Parser {
            tokens,
            filename,
            pos: 0,
            nodes: Vec::with_capacity(cap),
            index: Vec::with_capacity(cap),
            diags: Vec::new(),
            gui: false,
        }
    }

    /// Appends a node and returns its id.
    fn alloc_node(&mut self, node: Node) -> NodeId {
        let id = NodeId::new(self.nodes.len() as u32);
        self.nodes.push(node);
        id
    }

    /// The token kind at `pos + offset`, or [`TokenKind::Eof`] past the end.
    fn peek(&self, offset: usize) -> TokenKind {
        match self.tokens.get(self.pos + offset) {
            Some(t) => t.kind,
            None => TokenKind::Eof,
        }
    }

    /// Advances one token. (Return value is unused by callers that only need the
    /// side effect; callers that need the token read it before advancing.)
    fn advance(&mut self) {
        if self.pos < self.tokens.len() {
            self.pos += 1;
        }
    }

    /// The byte offset to attribute a diagnostic to at the current position.
    fn current_offset(&self) -> u32 {
        if self.pos < self.tokens.len() {
            self.tokens[self.pos].range.start
        } else if let Some(last) = self.tokens.last() {
            last.range.end
        } else {
            0
        }
    }

    fn add_diag(&mut self, offset: u32, severity: Severity, message: String) {
        self.diags.push(Diagnostic {
            filename: Arc::clone(&self.filename),
            offset,
            message,
            severity,
        });
    }

    /// Skips tokens until a safe resumption point: a closing `}`, the start of a
    /// plausible new item, or EOF. Does NOT consume the token it stops at.
    fn synchronize(&mut self) {
        loop {
            match self.peek(0) {
                TokenKind::Eof | TokenKind::RBrace => return,
                t0 if is_atom(t0) => {
                    let t1 = self.peek(1);
                    if is_operator(t1) || t1 == TokenKind::LBrace {
                        return;
                    }
                    self.advance();
                }
                _ => self.advance(),
            }
        }
    }

    /// Looks past a key (skipping `:`/`.` chain pairs) to see if an operator
    /// follows — i.e. whether this item is a field.
    fn peek_op_after_key(&self) -> bool {
        let mut i = 1;
        loop {
            let tag = self.peek(i);
            if tag == TokenKind::Colon || tag == TokenKind::Dot {
                i += 2;
                continue;
            }
            return is_operator(tag);
        }
    }

    // ── Parsing ─────────────────────────────────────────────────────────────

    /// Parses top-level items and returns the root node id (always `0`).
    fn parse_file(&mut self) -> NodeId {
        let root = self.alloc_node(blank(NodeKind::File));
        let mut child_ids: Vec<NodeId> = Vec::new();

        while self.peek(0) != TokenKind::Eof {
            if let Some(id) = self.parse_item() {
                child_ids.push(id);
            }
        }

        let start = self.index.len() as u32;
        self.index.extend_from_slice(&child_ids);
        let end = self.index.len() as u32;
        let root_node = &mut self.nodes[root.index()];
        root_node.child_start = start;
        root_node.child_end = end;
        root
    }

    /// Parses one item, or `None` for a skipped token.
    fn parse_item(&mut self) -> Option<NodeId> {
        let t0 = self.peek(0);
        if t0 == TokenKind::Eof {
            return None;
        }
        if t0 == TokenKind::RBrace || t0 == TokenKind::RBracket {
            self.advance();
            return None;
        }
        if t0 != TokenKind::Minus && is_atom(t0) && self.peek_op_after_key() {
            return self.parse_field();
        }
        self.parse_value(0)
    }

    /// Parses a `key OP value` field, or `None` on error.
    fn parse_field(&mut self) -> Option<NodeId> {
        let key_start = self.tokens[self.pos].range.start;
        let mut key_end = self.tokens[self.pos].range.end;
        self.advance();
        while self.peek(0) == TokenKind::Colon || self.peek(0) == TokenKind::Dot {
            self.advance(); // connector
            if self.pos < self.tokens.len() {
                key_end = self.tokens[self.pos].range.end;
                self.advance();
            }
        }

        if !is_operator(self.peek(0)) {
            let off = self.current_offset();
            let got = self.peek(0);
            self.add_diag(
                off,
                Severity::Error,
                format!("expected operator, got {}", got.as_str()),
            );
            self.synchronize();
            return None;
        }
        let op_tok = self.tokens[self.pos];
        self.advance();

        let key_id = self.alloc_node(Node {
            kind: NodeKind::Scalar,
            range: TextRange::new(key_start, key_end),
            operator: TokenKind::Invalid,
            child_start: 0,
            child_end: 0,
        });

        // scripted_trigger/effect call sites pass a bare comparator as a macro
        // argument, e.g. `OPERATOR = <=`. After a single '=', accept a comparator
        // token as the value. (A non-'=' first operator stays a real error.)
        let val_id = if op_tok.kind == TokenKind::Equal && is_operator(self.peek(0)) {
            let cmp_tok = self.tokens[self.pos];
            self.advance();
            self.alloc_node(Node {
                kind: NodeKind::Scalar,
                range: cmp_tok.range,
                operator: TokenKind::Invalid,
                child_start: 0,
                child_end: 0,
            })
        } else {
            self.parse_value(0)?
        };

        let idx_start = self.index.len() as u32;
        self.index.push(key_id);
        self.index.push(val_id);

        Some(self.alloc_node(Node {
            kind: NodeKind::Field,
            range: TextRange::new(key_start, key_end),
            operator: op_tok.kind,
            child_start: idx_start,
            child_end: idx_start + 2,
        }))
    }

    /// Parses block items until `}` or EOF. `lbrace` is the opening brace token,
    /// used for the unclosed-block diagnostic offset.
    fn parse_block_items(&mut self, lbrace: Token) -> Vec<NodeId> {
        let mut items: Vec<NodeId> = Vec::new();
        while self.peek(0) != TokenKind::RBrace && self.peek(0) != TokenKind::Eof {
            if let Some(id) = self.parse_item() {
                items.push(id);
            }
        }
        if self.peek(0) == TokenKind::RBrace {
            self.advance();
        } else {
            // EOF without a closing brace. An inner unclosed block may have eaten
            // the '}' meant for this block, so the real mistake may be nested.
            self.add_diag(
                lbrace.range.start,
                Severity::Error,
                "unclosed block (missing '}'; an inner block may have stolen the closing brace)"
                    .to_string(),
            );
        }
        items
    }

    /// Parses a value, or `None` on error.
    fn parse_value(&mut self, min_bp: i32) -> Option<NodeId> {
        // Unary minus.
        if self.peek(0) == TokenKind::Minus {
            let start = self.tokens[self.pos].range.start;
            self.advance();
            if self.pos >= self.tokens.len() {
                let off = self.current_offset();
                self.add_diag(off, Severity::Error, "unexpected EOF after '-'".to_string());
                return None;
            }
            let mut end = self.tokens[self.pos].range.end;
            self.advance();
            while binding_power(self.peek(0)) > min_bp {
                self.advance(); // connector
                if self.pos < self.tokens.len() {
                    end = self.tokens[self.pos].range.end;
                    self.advance();
                }
            }
            return Some(self.alloc_node(Node {
                kind: NodeKind::Scalar,
                range: TextRange::new(start, end),
                operator: TokenKind::Invalid,
                child_start: 0,
                child_end: 0,
            }));
        }

        // Tagged block.
        if self.peek(0) == TokenKind::Identifier && self.peek(1) == TokenKind::LBrace {
            let tag_tok = self.tokens[self.pos];
            self.advance(); // tag
            let lbrace = self.tokens[self.pos];
            self.advance(); // consume '{'
            let items = self.parse_block_items(lbrace);
            let idx_start = self.index.len() as u32;
            self.index.extend_from_slice(&items);
            let idx_end = self.index.len() as u32;
            return Some(self.alloc_node(Node {
                kind: NodeKind::TaggedBlock,
                range: tag_tok.range,
                operator: TokenKind::Invalid,
                child_start: idx_start,
                child_end: idx_end,
            }));
        }

        // Plain block.
        if self.peek(0) == TokenKind::LBrace {
            let lbrace = self.tokens[self.pos];
            self.advance(); // consume '{'
            let items = self.parse_block_items(lbrace);
            let idx_start = self.index.len() as u32;
            self.index.extend_from_slice(&items);
            let idx_end = self.index.len() as u32;
            return Some(self.alloc_node(Node {
                kind: NodeKind::Block,
                range: TextRange::new(0, 0),
                operator: TokenKind::Invalid,
                child_start: idx_start,
                child_end: idx_end,
            }));
        }

        // Interface dialect: `[Datafunction.Chain( … )]` as one scalar value
        // covering the whole bracketed text. Brackets do not nest in the
        // datafunction language, so scan to the first `]`.
        if self.gui && self.peek(0) == TokenKind::LBracket {
            let start = self.tokens[self.pos].range.start;
            let mut end = self.tokens[self.pos].range.end;
            self.advance();
            while self.peek(0) != TokenKind::RBracket && self.peek(0) != TokenKind::Eof {
                end = self.tokens[self.pos].range.end;
                self.advance();
            }
            if self.peek(0) == TokenKind::RBracket {
                end = self.tokens[self.pos].range.end;
                self.advance();
            } else {
                self.add_diag(
                    start,
                    Severity::Error,
                    "unclosed datafunction (missing ']')".to_string(),
                );
            }
            return Some(self.alloc_node(Node {
                kind: NodeKind::Scalar,
                range: TextRange::new(start, end),
                operator: TokenKind::Invalid,
                child_start: 0,
                child_end: 0,
            }));
        }

        // Atom + optional scope-chain infix.
        if !is_atom(self.peek(0)) {
            let off = self.current_offset();
            let got = self.peek(0);
            self.add_diag(
                off,
                Severity::Error,
                format!("expected value, got {}", got.as_str()),
            );
            self.synchronize();
            return None;
        }
        let start = self.tokens[self.pos].range.start;
        let mut end = self.tokens[self.pos].range.end;
        self.advance();
        while binding_power(self.peek(0)) > min_bp {
            self.advance(); // connector
            if self.pos < self.tokens.len() {
                end = self.tokens[self.pos].range.end;
                self.advance();
            }
        }
        Some(self.alloc_node(Node {
            kind: NodeKind::Scalar,
            range: TextRange::new(start, end),
            operator: TokenKind::Invalid,
            child_start: 0,
            child_end: 0,
        }))
    }
}

/// Whether `tag` can begin an atom/value.
fn is_atom(tag: TokenKind) -> bool {
    matches!(
        tag,
        TokenKind::Identifier
            | TokenKind::LiteralNumber
            | TokenKind::LiteralString
            | TokenKind::LiteralBoolean
            | TokenKind::LiteralDate
            | TokenKind::Minus
            | TokenKind::MacroParam
            | TokenKind::ScriptValue
            | TokenKind::ScriptMath
    )
}

/// Whether `tag` is an assignment/comparison operator.
fn is_operator(tag: TokenKind) -> bool {
    matches!(
        tag,
        TokenKind::Equal
            | TokenKind::EqualEqual
            | TokenKind::NotEqual
            | TokenKind::QuestionEqual
            | TokenKind::GreaterThan
            | TokenKind::GreaterEqual
            | TokenKind::LessThan
            | TokenKind::LessEqual
    )
}

/// Infix binding power for scope-chain connectors (`.`, `:`, `|` → 80).
fn binding_power(tag: TokenKind) -> i32 {
    match tag {
        TokenKind::Colon | TokenKind::Dot | TokenKind::Pipe => 80,
        _ => 0,
    }
}
